rqd/rqd/rqcore.py

```diff
@@ -960,6 +960,93 @@ def __cleanup(self):
+
+        self.__createEnvVariables()
+        self.__writeHeader()
+
+        # TODO: Better handling if env vars are not set or not able to connect to the systemd-daemon
+        self.frameEnv['DBUS_SESSION_BUS_ADDRESS'] = os.environ['DBUS_SESSION_BUS_ADDRESS']
+        systemdUnitName = f"rqd-{frameInfo.frameId}.service"
+
+        def journalReaderThread():
+            rules = (
+                Rule("_SYSTEMD_USER_UNIT", systemdUnitName)
+            )
+            journalReader = JournalReader()
+            journalReader.open(JournalOpenMode.CURRENT_USER)
+            journalReader.seek_tail()
+            journalReader.add_filter(rules)
+            while True:
+                status = journalReader.wait(1)  # It will skip if there is no event for 1 second
+                if status == JournalEvent.NOP and frameInfo.forkedCommand.returncode is not None:
+                    break
+                for record in journalReader:
+                    self.rqlog.write(f"{record.data['MESSAGE']}\n",
+                                     prependTimestamp=rqd.rqconstants.RQD_PREPEND_TIMESTAMP)
+
+        readerThread = threading.Thread(target=journalReaderThread)
+        readerThread.start()
+
+        tempCommand = [
+            "systemd-run",
+            "--user",
+            "--wait",
+            "--collect",
+            f"--unit={systemdUnitName}"
+        ]
+
+        if 'CPU_LIST' in runFrame.attributes:
+            tempCommand.append("--property=CPUAffinity=runFrame.attributes['CPU_LIST']")
+
+        if self.rqCore.machine.isDesktop():
+            tempCommand.append("--property=Nice=10")
+
+        if rqd.rqconstants.RQD_BECOME_JOB_USER:
+            tempCommand.append(f"--property=User={runFrame.user_name}")
+
+        tempCommand += ['/bin/sh']
+        rqd.rqutil.permissionsHigh()
+        try:
+            tempCommand += [self._createCommandFile(runFrame.command)]
+            # pylint: disable=subprocess-popen-preexec-fn,consider-using-with
+            frameInfo.forkedCommand = subprocess.Popen(tempCommand,
+                                                       env=self.frameEnv,
+                                                       cwd=self.rqCore.machine.getTempPath(),
+                                                       stdout=subprocess.DEVNULL,
+                                                       stderr=subprocess.DEVNULL,
+                                                       close_fds=True)
+        finally:
+            rqd.rqutil.permissionsLow()
+
+        frameInfo.pid = runFrame.pid = frameInfo.forkedCommand.pid
+
+        if not self.rqCore.updateRssThread.is_alive():
+            self.rqCore.updateRssThread = threading.Timer(rqd.rqconstants.RSS_UPDATE_INTERVAL,
+                                                          self.rqCore.updateRss)
+            self.rqCore.updateRssThread.start()
+
+        returncode = frameInfo.forkedCommand.wait()
+        readerThread.join()
```

Comment:
If an exception happens between the reader thread and here, you might have a thread leaked. Using a try/finally might add extra protection.